src/rewards/mod.rs

```diff
@@ -143,102 +162,194 @@ impl<'a> RewardsMonitor<'a> {
+        let epoch_durations = (current_epoch - MAX_EPOCH_LOOKBACK + 1..=current_epoch)
+            .map(|epoch| Ok((epoch, self.epoch_duration_days(epoch, current_epoch_info)?)))
+            .collect::<anyhow::Result<BTreeMap<_, _>>>()?;
+        let duration_max_epoch_lookback: f64 = epoch_durations.values().sum();
+
+        let mut voter_apys = HashMap::new();
+        for (voter, epoch_apys) in voter_epoch_apys {
+            let mut total_apy = 0.0;
+            for (epoch, duration) in &epoch_durations {
+                let apy = *epoch_apys.get(epoch).unwrap_or(&0.0);
+                total_apy += apy * duration;
             }
+            let average_apy = total_apy / duration_max_epoch_lookback;
+            let current_apy = *epoch_apys.get(&current_epoch).unwrap_or(&0.0);
+            voter_apys.insert(
+                voter,
+                VoterApy {
+                    current_apy,
+                    average_apy,
+                },
+            );
+        }
+        Ok(voter_apys)
+    }
 
-            // Extend accounts
-            accounts.extend(pka);
+    fn epoch_duration_days(&self, epoch: Epoch, epoch_info: &EpochInfo) -> anyhow::Result<f64> {
+        // TODO: Use some extrapolation magic to find out the epoch length if it's the current epoch
+        // If it's the current epoch we can't use the next epoch.
+        if epoch == epoch_info.epoch {
+            return Ok(3.0);
         }
 
-        todo!("missing staking apy calculation using multiple epochs")
+        if let Some(length) = self.cache.get_epoch_length(epoch)? {
+            Ok(length)
+        } else {
+            debug!("Finding epoch {}", epoch);
+            let days_in_epoch = {
+                let start_timestamp = self
+                    .client
+                    .get_blocks(
+                        epoch * epoch_info.slots_in_epoch,
+                        Some((epoch * epoch_info.slots_in_epoch) + SLOT_OFFSET),
+                    )?
+                    .get(0)
+                    .cloned()
+                    .map(|slot| self.client.get_block(slot))
+                    .transpose()?
+                    .and_then(|x| x.block_time);
+
+                let end_timestamp = self
+                    .client
+                    .get_blocks(
+                        (epoch + 1) * epoch_info.slots_in_epoch,
+                        Some(((epoch + 1) * epoch_info.slots_in_epoch) + SLOT_OFFSET),
+                    )?
+                    .get(0)
+                    .cloned()
+                    .map(|slot| self.client.get_block(slot))
+                    .transpose()?
+                    .and_then(|x| x.block_time);
+
+                if let (Some(start_timestamp), Some(end_timestamp)) =
+                    (start_timestamp, end_timestamp)
+                {
+                    (end_timestamp - start_timestamp) as f64 / 86_400_f64
+                } else {
+                    3.0
```

Comment:
This prominent number deserves a `const` :)

Reply:
Fixed in 912c8646